runtime/compiler/control/CompilationRuntime.hpp

```diff
@@ -457,17 +457,31 @@ class CompilationInfo
    static JITServer::ServerStream *getStream();
 #endif /* defined(J9VM_OPT_JITSERVER) */
    static bool isInterpreted(J9Method *method) { return !isCompiled(method); }
-   static bool isCompiled(J9Method *method)
+
+   /**
+    * @brief Determines if a J9Method is compiled
+    *
+    * @param method pointer to the J9Method
+    *
+    * @return The start PC if compiled, NULL otherwise
+    */
+   static void* isCompiled(J9Method *method)
       {
 #if defined(J9VM_OPT_JITSERVER)
       if (auto stream = getStream())
          {
          stream->write(JITServer::MessageType::CompInfo_isCompiled, method);
-         return std::get<0>(stream->read<bool>());
+         return std::get<0>(stream->read<void *>());
          }
 #endif /* defined(J9VM_OPT_JITSERVER) */
-      return (((uintptr_t)method->extra) & J9_STARTPC_NOT_TRANSLATED) == 0;
+
+      /* Read extra field only once */
+      void *extra = method->extra;
+
+      /* Return extra field if compiled, NULL otherwise */
+      return ((uintptr_t)extra & J9_STARTPC_NOT_TRANSLATED) == 0 ? extra : NULL;
```

Comment:
I'm not convinced the compiler has to respect this and only read the field once.  Maybe I should say that more strongly - there's no guarantee the compiler will read `method->extra` only once here.

Reply:
I'm confused. The code as written cannot see different values for the two uses of the local variable "extra". The compiler isn't allowed to change that...

Reply:
Couldn't the compiler see that the definition of `extra` comes from `method->extra` and just choose to use that since there's been no store to the local variable `extra`?

Reply:
As @dsouzai says, the compiler doesn't have to create a local for `method->extra` (privatize) and is free to continue to fetch from the `method`.  We've seen occurrences of this kind of behaviour on some platforms in the past, especially when code like this is inlined into other methods

Reply:
ok, yup, must be volatile *and* we need this change then

Reply:
Would `void * volatile extra = method->extra;` prevent the compiler from using `method->extra` instead of the local `extra`?
I am thinking that adding the `volatile` keyword changes the type, so the compiler may not be able to use `method->extra` and `extra` interchangeably.

Reply:
I still find it mildly disturbing that a C compiler would de-privatize an access to a global variable, but I can also see the line of reasoning that led the optimization folks to find that behaviour desirable, relying on the volatile keyword to indicate that the value can change underfoot.

That got me thinking though: do any other variables that can be read or written to without a protective mutex need to be similarly marked as volatile? stackOverflowMark in the VM thread maybe?

Reply:
I would think that the struct field being volatile ensures the compiler will use the local copy.

Reply:
I agree. I was wondering if there are any other fields that the JIT accesses directly that could be concurrently updated that should also be marked as volatile (basically, are we sure that this one is the only instance of the problem?).

Reply:
Most of the fields that are concurrently updated are mutexxed, but I'm sure there are a few. J9VMThread.stackOverflowMark comes to mind immediately (since this caused problems on ZOS XLC). All of the RAM constant pool structs should probably be volatile (Jason fixed one in the last year that was causing MSVC to corrupt the field), though they are a somewhat special case in that there are only two possible states (multithreaded resolution always results in the same final values).